### ISSUE TITLE
Image Export plugin crashes on batchEdgesDrawing: true

### ISSUE THREAD

**Issue by emergingdzns** [Linkurious/linkurious.js#181]
Following the directions and example:
```sigma.plugins.image(s, s.renderers[0], {download:true});```

But when the image export is attempted, I get the following error message in the console:

```InvalidStateError: An attempt was made to use an object that is not, or is no longer, usable
_canvasContext.drawImage(canvas, 0, 0);``` - referring to line 190 on the plugin js file.

I can't figure out what object it's referring to. Any ideas?

**Comment by sheymann** [comment]
Can you check if `canvas` is not null using the debugger?

**Comment by emergingdzns** [comment]
Yes. I console logged "canvas", "_canvas" and "_canvasContext" and none are blank. Here is the output:
canvas:
```<canvas class="sigma-scene" style="position: absolute; width: 0px; height: 0px;" width="0" height="0px">```

_canvas:
```<canvas width="0" height="0">```

_canvasContext:
``` CanvasRenderingContext2D { canvas=canvas,  globalAlpha=1,  globalCompositeOperation="source-over",  more...}```

I'm not sure if it helps or matters, but if I inspect elements, I can see the image-container div and the style is `width:NaNpx;height:NaNpx;`. I tried setting it to the number values of the visible canvas for the graph using firebug, but when I click the button to make the image, it resets the values and adds 3 more elements.

**Comment by sheymann** [comment]
There is an issue with the image generation. What happens for `sigma.plugins.image(s, s.renderers[0], {download:false});` ?

**Comment by emergingdzns** [comment]
I get the same error message.

**Comment by sheymann** [comment]
Does it happen on all web browsers?

**Comment by sheymann** [comment]
Notice that Safari 8 is buggy and crashes on canvas rendering, see http://stackoverflow.com/questions/27593998/safari-8-freezes-while-using-canvas

**Comment by emergingdzns** [comment]
Thanks. Unfortunately, I'm experiencing the error message in Firefox, not Safari. I did just load it up in Chrome and the error I'm getting there when performing the same image export action is different:
```
Uncaught TypeError: Cannot read property 'edges' of undefined
job @ sigma.renderers.canvas.js:207
_executeFirstJob @ conrad.js:271
_loop @ conrad.js:363
```
I appreciate the help. I'm a bit confused by the code. I'm usually a pretty good debugger but this stuff has got me stumped.

**Comment by sheymann** [comment]
Call the method with the `size` option set:
`sigma.plugins.image(s, s.renderers[0], {download:true, size:1000});`

**Comment by emergingdzns** [comment]
Ok. That's different... Got past that previous error and now gives this in Firefox:
```
TypeError: this.contexts is undefined
tempGCO = this.contexts.edges.globalCompositeOperation;
```
Chrome however gives the same error about unable to read the property 'edges'.

Safari gives a similar error as Firefox:
```
[Error] TypeError: undefined is not an object (evaluating 'this.contexts.edges')
	job (sigma.renderers.canvas.js, line 207)
	(anonymous function) ([native code], line 0)
	_executeFirstJob (conrad.js, line 271)
	_loop (conrad.js, line 363)
```

These to me seem to indicate that it has gotten past the image plugin and is now having issues in the core somewhere reading the edges. I have a working graph though so I'm not sure what's breaking it. I'll try to see if I can put together a jsfiddle of this graph, but that will take some time.

**Comment by emergingdzns** [comment]
@sheymann I did some debugging and found a work-around. Perhaps I'm doing something wrong that makes this necessary but I'm not sure. Anyway, since we were getting past that error on line 190 of the image plugin, I went into the sigma.renderers.canvas.js file and added some console logs.

At line 211 (immediately following the `job = function() {` line, I added:
```
console.log(id); // refers to line 201
console.log(this.contexts);
```
When I reload the page, I get a bunch of lines like this in my console:
```
edges_1
Object { edges=CanvasRenderingContext2D,  scene=CanvasRenderingContext2D,  nodes=CanvasRenderingContext2D,  more...}
```
No errors on load and all I see is edges_1 followed by the object line.

However, when I click the button to create the image, I get a single set of the above followed by a single line saying `edges_2` and then the TypeError block about this.contexts is undefined.

I followed the code up and found that if I go back to the image plugin file and change line 182 from `batchEdgesDrawing: true` to `batchEdgesDrawing: false` I now get an image export working.

I don't understand what that batchEdgesDrawing setting is for really, but it fixed the problem. I don't want to leave the plugin modified though, but I don't see a setting for changing this in the call to the plugin.

Hopefully this helps you figure out what might be happening.
